File: 655-print-binary-tree/655-print-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def printTree(self, root: Optional[TreeNode]) -> List[List[str]]:
        d = self.depth(root)
        ans = [['']*(2**d-1) for _ in range(d)]
        
        def dfs(node, row, pos):
            ans[row][pos] = str(node.val)
            if node.left:
                dfs(node.left, row+1, pos-2**(d-row-2))
            if node.right:
                dfs(node.right, row+1, pos+2**(d-row-2))
                
        dfs(root, 0, 2**(d-1)-1)
        
        return ans
        
        
    def depth(self, node):
        if not node:
            return 0
        return 1 + max(self.depth(node.left), self.depth(node.right))
        

File: 655-print-binary-tree/655-print-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[][]}
 */
var printTree = function (root) {
  const rows = getMaxDepth(root);
  const ans = Array.from({ length: rows }, () =>
    new Array(2 ** rows - 1).fill("")
  );

  const fillNum = (node, row, pos) => {
    ans[row++][pos] += node.val;

    if (node.left) fillNum(node.left, row, pos - 2 ** (rows - 1 - row));
    if (node.right) fillNum(node.right, row, pos + 2 ** (rows - 1 - row));
  };

  fillNum(root, 0, 2 ** (rows - 1) - 1);

  return ans;
};

const getMaxDepth = (node) => {
  if (!node) return 0;

  return 1 + Math.max(getMaxDepth(node.left), getMaxDepth(node.right));
};
